### PR TITLE
[TRACKING]  xtimer/xtimer.c: xtimer_mutex_lock_timeout fix and improvements

### PR DESCRIPTION
**Pull request by JulianHolzwarth**
This pr tracks the fix for `xtimer/xtimer.c: xtimer_mutex_lock_timeout`
This pr includes two throwaway test commits.

### Contribution description
This pr fixes Bugs in `xtimer.c: xtimer_mutex_lock_timeout()` and improves it. It addresses concurrency issues. The timer can trigger at a different time than expected.
Before there were problems when the timer spins. This happens when the timeout is less than `XTIMER_BACKOFF`. 

This pr also creates a new static function called `_mutex_remove_thread_from_waiting_queue`. The function removes a thread from a mutex waiting queue, when the thread waiting for the mutex. This was already used inside of the `_mutex_timeout` function. It is needed for the `xtimer_mutex_lock_timeout` function.
This function should go into` core/mutex` in the future because it modifies queue of the mutex_t struct. (in mutex_t struct comment: `@brief Mutex structure. Must never be modified by the user.` and in comment of queue `@brief   The process waiting queue of the mutex. **Must never be changed by the user.**`)

This pr also fixes `core/mutex.c:_mutex_lock()` there was a problem when it gets interrupted before disabling interrupt. This is fixed by giving it a pointer to blocking.

This pr also has tests and throwaway commits to show the bug and fix.

The struct for the `xtimer_mutex_lock_timeout` function called `mutex_thread_t` was changed to fix the function and name changes for better understanding.

This pull request is split into smaller parts:
 - [x] tests/xtimer_mutex_lock_timeout: add simple case test #11679 
 - [x] sys/xtimer/xtimer.c: _mutex_timeout() cleanup #11807
 - [x] sys/xtimer/xtimer.c: _mutex_timeout() bug fix #11992
 - [x] tests/xtimer_mutex_lock_timeout: New test #12008
 - [x] xtimer/xtimer.c:_mutex_timeout() improved #13185
 - [x] xtimer/xtimer.c: xtimer_mutex_lock_timeout fix with short timeout #13199
 

### Testing procedure

`BOARD=native make -C tests/xtimer_mutex_lock timeout/ flash test`
it outputs:
```
...
> mutex_timeout_long_unlocked
mutex_timeout_long_unlocked
starting test: xtimer mutex lock timeout
OK

> mutex_timeout_long_locked
mutex_timeout_long_locked
starting test: xtimer mutex lock timeout
OK

> mutex_timeout_long_locked_low
mutex_timeout_long_locked_low
starting test: xtimer mutex lock timeout with thread
threads = 2
THREAD low prio: start
MAIN THREAD: calling xtimer_mutex_lock_timeout
OK
threads = 3
MAIN THREAD: waiting for created thread to end
THREAD low prio: exiting low
threads = 2

> mutex_timeout_short_locked
mutex_timeout_short_locked
starting test: xtimer mutex lock timeout with short timeout and locked mutex
OK

> mutex_timeout_short_unlocked
mutex_timeout_short_unlocked
starting test: xtimer mutex lock timeout with short timeout and unlocked mutex
OK

>
```

Without the fixes this would fail.

The remove me commits show other bugs.
Go to the commit before "REMOVE ME" commit gets reverted.
Test will work.

Now go to the "REMOVE ME" commit. 
Test will fail.

To test the bug in mutex.c:_mutex_lock revert to the second "REMOVE ME" commit with the commit message 

> REMOVE ME
    
>  test commit to show bug in _mutex_lock
>  when interupted before irq_disable

### Issues/PRs references

#11679: first tests
#11807: sys/xtimer/xtimer.c: _mutex_timeout() clean ups
#11992: sys/xtimer/xtimer.c: _mutex_timeout() bug fix
#12008: tests/xtimer_mutex_lock_timeout: New test
#13185: xtimer/xtimer.c:_mutex_timeout() improved 
#13199: xtimer/xtimer.c: xtimer_mutex_lock_timeout fix with short timeout 

needed for: 
    - #11977: xtimer/xtimer.c: xtimer_rmutex_lock_timeout
    - #11485: [TRACKING] FreeRTOS Adaption Layer

### CODE REVIEW

**Comment by JulianHolzwarth** [comment]
rebase because #11679 got merged

File: sys/xtimer/xtimer.c [hunk 1]
```diff
@@ -237,34 +244,66 @@ int _xtimer_msg_receive_timeout(msg_t *msg, uint32_t timeout_ticks)
...
+        list_node_t *node = list_remove(&mutex->queue, (list_node_t *)&thread->rq_entry);
+        /* if thread was removed from the list */
+        if (node != NULL) {
+            DEBUG("thread: %"PRIkernel_pid" removed from list\n", (container_of((clist_node_t *)node, thread_t, rq_entry)->pid));
+            if (mutex->queue.next == NULL) {
+                mutex->queue.next = MUTEX_LOCKED;
+            }
+            sched_set_status(thread, STATUS_PENDING);
+            irq_restore(irqstate);
+            thread_yield_higher();
```
Comment (cladmi) [inline sys/xtimer/xtimer.c]: It looks like it is `sched_switch` that may need to be used here instead to be working everywhere.
But from the implementation, `thread_yield_higher` in `cpu/atmega_common/thread_arch.c`, does the check for `is_in_irq`…

I will do a dedicated test for this.

Comment (cladmi) [reply sys/xtimer/xtimer.c]: https://github.com/RIOT-OS/RIOT/pull/11759

Comment (cladmi) [reply sys/xtimer/xtimer.c]: Implementation in the CPU is wrong, this code is correct https://github.com/RIOT-OS/RIOT/pull/11759#issuecomment-506760423

Comment (JulianHolzwarth) [reply sys/xtimer/xtimer.c]: I will change it to `sched_switch` for now until #11759 is solved/discussed/closed.

Comment (JulianHolzwarth) [reply sys/xtimer/xtimer.c]: done in sys/xtimer/xtimer.c: _mutex_timeout() clean ups #11807

**Comment by JulianHolzwarth** [comment]
rebased on master

**Comment by JulianHolzwarth** [comment]
This pr still needs some work. (rebase is not enough)
I will update it next week.

**Comment by JulianHolzwarth** [comment]
rebased on master

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner done. Removed debug msg.

File: core/include/mutex.h [hunk 2]
```diff
@@ -100,7 +101,8 @@ int _mutex_lock(mutex_t *mutex, int blocking);
  */
 static inline int mutex_trylock(mutex_t *mutex)
 {
-    return _mutex_lock(mutex, 0);
+    uint8_t blocking = 0;
```
Comment (MichelRottleuthner) [inline core/include/mutex.h]: This should be declared as volatile explicitly.

File: core/include/mutex.h [hunk 3]
```diff
@@ -110,7 +112,8 @@ static inline int mutex_trylock(mutex_t *mutex)
  */
 static inline void mutex_lock(mutex_t *mutex)
 {
-    _mutex_lock(mutex, 1);
+    uint8_t blocking = 1;
```
Comment (MichelRottleuthner) [inline core/include/mutex.h]: same here

File: sys/xtimer/xtimer.c [hunk 4]
```diff
@@ -247,6 +247,7 @@ static void _mutex_timeout(void *arg)
     unsigned int irqstate = irq_disable();
 
     mutex_thread_t *mt = (mutex_thread_t *)arg;
+
```
Comment (MichelRottleuthner) [inline sys/xtimer/xtimer.c]: unrelated

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner can I just rebase after the changes?

**Comment by MichelRottleuthner** [comment]
sure

**Comment by JulianHolzwarth** [comment]
done
